Source/UI/MLCardFormField/Mask/MLCardFormCustomMask.swift

```diff
@@ -6,118 +6,78 @@ import Foundation
  *
  *  usage:
  *   func textField(textField: UITextField, shouldChangeCharactersInRange range: NSRange, replacementString string: String) -> Bool {
- *      self.text = customMask!.formatStringWithRange(range, string: string)
+ *      textField.text = customMask.shouldChangeCharactersIn(range, with: string)
  *      return false
  *   }
  *
  *  Use $ for digits ($$-$$$$)
  *  Use * for characters [a-zA-Z] (**-****)
  */
 
-class MLCardFormCustomMask {
+final class MLCardFormCustomMask {
+    let mask: String
+    private(set) var value: String = ""
+    var unmaskedText: String { removeMask(to: value) }
     
-    var finalText: String? = String()
-    var bufferText: String? = String()
-
-    private var mask: String?
-    
-    public init(mask: String = String()){
+    init(mask: String) {
         self.mask = mask
     }
-
-    public func formatString(string: String) -> String {
-        
-        self.finalText = string
-        
-        self.bufferText = string
-        
-        self.finalText?.applyMask(self.mask)
-        
-        return self.finalText ?? String()
-    }
-
-    public func formatStringWithRange(range: NSRange, string: String) -> String {
-        
-        guard !string.isEmpty else {
-            
-            if range.upperBound <= finalText?.count ?? 0 {
-                finalText?.removeLast()
-            }
-            
-            if range.upperBound <= bufferText?.count ?? 0 {
-                bufferText?.removeLast()
-            }
-            
-            return finalText ?? String()
-        }
-        
-        finalText?.appendCharWithMask(Character(string), mask: self.mask)
-        
-        bufferText?.append(string)
-        
-        return finalText ?? String()
+    
+    func clear() {
+        value.removeAll()
     }
-}
-
-internal extension String {
```

Comment:
`MLCardCustomMask` follows the SRP because it will only change if the logic for the mask need to change.
The functionality to apply a mask is not needed to be available to every string and as an example this code was not used anywhere else in the lib. 
If we wanted to expose a way to apply a mask (with the given rules) to every string, then we can create a free function like `func applyMask(_ mask: String, to string: String) -> String` but It would lack of context if then we want to create a free function to remove that given mask `func removeMask(_ mask: String, from string: String) -> String` so that would bring us into create a name space that would give context about the type of mask supported... that context is already given by the class `MLCardCustomMask` and it even have a wrong name since it is not only for cards but also for identifier numbers, CVV, expiration date and more.
It should be something like 

```swift
struct MaskFormatter {
    let mask: String
    func maskedString(from string: String) -> String
    func unmaskedString(from string: String) -> String
 }
```
and that would be all, then each textfield is responsible for the logic of 
```swift
    public func textField(_ textField: UITextField,
                   shouldChangeCharactersIn range: NSRange,
                   replacementString string: String) -> Bool {
            let newString = (textField.text as NSString).replacingCharacters(in: range, with: string)
            textField.text = maskFormatter.maskedString(from: newString)
            return false
    }
```
because the Formatter should not be responsable of keeping the consistency of the range you want to replace, it is not part of its context/responsability.

Reply:
I will have to disagree on that. `ApplyMask` and `RemoveMask` are features related to the scope of the entity `String`, not to the specific object `MLCardFormCustomMask`. This object is an intermediary entity that applies the rules of masking, not the mask itself. By adding them to `MLCardFormCustomMask` we are specifying it in a single case, and transforming the `MLCardFormCustomMask' into a `God Class*`. 

We don't use it today in other parts of the code, but what if we will need in the future? In my personal opinion, this change breaks the OO.

Also the benefit of using extensions is that they are safer than derived or custom classes, because with then we can avoid a lot of problems related to casts, options, and the open possibility of unnecessary force unwrappings. We can also limit the range of the extension with "internal".